### ISSUE TITLE
CDRs module/Object description/CDR object/charging_periods: misleading

### ISSUE THREAD

**Issue by SoloVeniaASaludar** [ocpi/ocpi#370]
In CDRs module => Object description => CDR object => charging_periods we can read:

```
List of Charging Periods that make up this charging session. A
session consists of 1 or more periods, where each period has a
different relevant Tariff.
```

This paragraph seems misleading, two different charging periods could have same tariff.

**Comment by reinierl** [comment]
Good point. Will be fixed later this week.

**Comment by SoloVeniaASaludar** [comment]
Thinking a few more about it, could be we can say that an usual case will be all the charging periods in a CDR with the same tariff, being the difference the element of the tariff that applies, and the restriction that no two consecutive (in time) charging periods can exists in a CDR related to the same tariff element.

Could be interesting add a tariff element identifier to the charging period ?

**Comment by reinierl** [comment]
but it seems that multiple tariff elements _can_ apply at the same time in the same session, as long as they're used for different dimensions. The tariffs module documentation says this after all:

> When the list of Tariff Elements contains more than one Element with the same Tariff Dimension (ENERGY/FLAT/TIME etc.), than the first Tariff Element with that Dimension in the list with matching Tariff Restrictions will be used. Only one Tariff per Element type can be active at any point in time, but multiple Tariff Types can be active at once. IE you can have an ENERGY element and TIME element active at the same time, but only the first valid element of each.

So I propose to remove the "where each period has a different relevant Tariff" remark altogether. That will leave specification of tariff logic to the Tariffs module where it belongs.

**Comment by reinierl** [comment]
This kind of leaves me with the question how to find the tariffcomponents that apply at a time t in a session S...

Is it:

```
for all tariffs T:
    if T matches S at t:
        for all T's elements e:
            if e's restrictions match S at t:
                return all e's price components
```

or rather:

```
tariffs: for all tariffs T
    if T matches S at t:
        for dimension in { TIME, ENERGY, PARKING_TIME, FLAT }:
            elements: for all T's elements e:
                if e's restrictions match S at t:
                    applicable_price_components[dimension] = e.components[dimension]
                    break elements
        break tariffs
```

@RobertdeLeeuw-IHomer Do you know what was intended? And if it's the latter, why do we have the three-level tariff/element/component hierarchy when a two-level tariff/component hierarchy seems to suit the use better?

**Comment by reinierl** [comment]
OK, reading the examples in the module, I get the impression that:

 * price component lookup should be done using the latter algorithm
 * and the purpose of the three-level hierarchy is to not have to repeat the same restrictions for every price component that applies under the same conditions as other price components

**Comment by SoloVeniaASaludar** [comment]
> Is it:
> 
> ```
> for all tariffs T:
>     if T matches S at t:
>         for all T's elements e:
>             if e's restrictions match S at t:
>                 return all e's price components
> ```
> 

By example, if the tariff starts with an element with an unique price component of type "parking" and no restrictions, this is the result of the algorithm, without continue working remainder elements and tariffs for components "flat", "energy", ... . 


> or rather:
> 
> ```
> for all tariffs T
>     if T matches S at t:
>        applicable_price_components = {}
>         for dimension in [ TIME, ENERGY, PARKING_TIME, FLAT ]:
>             elements: for all T's elements e:
>                 if e's restrictions match S at t:
>                     applicable_price_components[dimension] = e.price_components[dimension]
>                     break elements
>         return applicable_price_components
> ```
> 

Note that if one restriction is fulfilled, the "break elements" finishes the analysis of next elements, even when the fulfilled restriction could be in an element that applies only to one of the tariff dimension.

Moreover, the return interrupts the analysis of the next tariff, only first tariff is used in this algorithm.

[ Note also no real implementation can apply an algorithm as these ones for each time "t" (every second?) ]

**Comment by reinierl** [comment]
> By example, if the tariff starts with an element with an unique price component of type "parking" and no restrictions, this is the result of this algorithm, without continue working remainder elements and tariffs for components "flat", "energy", ... .

Yep. So under this interpretation, if you wanted to charge for multiple dimensions at once, you would have to make sure they're in the same Tariff Element.

     
> Note that if one restriction is fulfilled, the "break elements" finishes the analysis of next elements, even when the fulfilled restriction could be in an element that applies only to one of the tariff dimensions.
> Moreover, the return interrupts the analysis of the next tariff, only first tariff is used in this algorithm.

The `return` and `break` are basically saying "the _first_ matching tariff" and "the _first_ matching element" in an imperative style. You're also right that actually writing code like this is impractical because it would have to be re-evaluated at every time step. I was more going for a logically exact description of which tariff components apply at any given point in time.
